src/core/qgstiledownloadmanager.cpp

```diff
@@ -197,6 +233,11 @@ QgsTileDownloadManagerReply *QgsTileDownloadManager::get( const QNetworkRequest
     QObject::connect( entry.objWorker, &QgsTileDownloadManagerReplyWorkerObject::finished, reply, &QgsTileDownloadManagerReply::requestFinished );  // should be queued connection
 
     addEntry( entry );
+
+    if ( isCached( request ) )
+    {
+      QTimer::singleShot( 0, entry.objWorker, &QgsTileDownloadManagerReplyWorkerObject::replyFinished );
+    }
```

Comment:
I did try that before and the removeEntry function ended up being called twice because of signalQueueModified or something. When I'm thinking about it now, the cached requests shouldn't be pushed into cache in the first place, I made a change according to this realization.